### DISCUSSION TITLE
is it possible to send keyboard events to the main form?

### DISCUSSION THREAD

**Discussion by wnunezc** [cefsharp/CefSharp#3496] [Q&A]
I'd like to avoid having to implement the IKeyboardHandler interface, to handle my keyboard events.
this is an example of what I have for keyboard shortcuts:

```c#
private void Form1_KeyDown(object sender, KeyEventArgs e)
        {
            if (e.Modifiers == Keys.Control)
            {
                switch (e.KeyCode)
                {
                    case Keys.F12:
                        //Comando 1
                        break;
                    case Keys.N:
                        //Comando 2 nueva ventana
                        break;
                }
            }
        }
```

and in this way I initialize the control for the browser:

```c#
public void InitializeChromium(string Url)
        {
            CefSettings settings = new CefSettings();
            Cefinit = Cef.Initialize(settings);
            BrowserControl = new ChromiumWebBrowser(Url)
            {
                Name = "WebBrowser",
                MenuHandler = new SearchContextMenuHandler(this),
                DownloadHandler = new DownloadHandler(download_img)
            };
            Controls.Add(BrowserControl);
        }
```
but once the control has the focus there is no way that the form receives the keyboard shortcuts for example: ctrl + N:
is there a way to redirect these events to the main form?
or best way:
**is there a keyboard event lock, how can it be removed ?**

**Comment by amaitland** [comment]
Integrate CEF into your apps existing message loop message loop see https://github.com/cefsharp/CefSharp/wiki/General-Usage#multithreadedmessageloop

Or implement IKeyboardHandler (for simple cases this is preferable)

> is there a keyboard event lock, how can it be removed ?

What exactly is a keyboard event lock?

**Reply by wnunezc** [reply]
> What exactly is a keyboard event lock?

is when an event works correctly when the focus is on the form, when clicking on a control (browser control / Cefsharp), the event catcher stops working ... at that moment there is stoped, the form keydown stops function.

I have tried it:
1. create the listener for the keydown event in the form events.
2. create the browser control and add it to the form.
3. run the application
4. without clicking, press the combination of keys: result works.
5. click on the loaded page, press the previous combination of keys: result does NOT work.

**Reply by amaitland** [reply]
By default CEF uses a different thread to run it's message loop, keyboard event's won't bubble up to the parent form. This is expected. Implement IKeyboardHandler or integrate CEF into your apps existing message loop as described in https://github.com/cefsharp/CefSharp/wiki/General-Usage#multithreadedmessageloop

**Reply by wnunezc** [reply]
I have tried to implement IKeyboardHandler, but I cannot catch the key combination events, as simple as a Ctrl + F12, using my event capture code, I have tried to follow this example:

https://github.com/cefsharp/CefSharp/blob/master/CefSharp.WinForms.Example/Handlers/KeyboardHandler.cs

I do not have access to: `object sender`, `KeyEventArgs e` and I can't find in the documentation where it says how to get these data / objects.

**Reply by amaitland** [reply]
> I do not have access to: object sender, KeyEventArgs e and I can't find in the documentation where it says how to get these data / objects.

Your implying it's a WinForms event which is incorrect. You'll need to parse the low level key messages, check the modifier and windows key code. 

https://docs.microsoft.com/en-us/windows/win32/inputdev/virtual-key-codes

**Reply by wnunezc** [reply]
ok, i think this will take a long time. :(

**Reply by amaitland** [reply]
You can cast IWebBrowser to ChromiumWebBrowser and access the WinForms  UI, making sure to marshal onto the WinForms UI thread to using BeginInvoke. 

Past that it's just a matter of some simple if checks or a switch statement. Might look roughly something like the following 

```c#
browser.KeyboardHandler = new KeyboardHandler();
public class KeyboardHandler : IKeyboardHandler
    {
        public bool OnKeyEvent(IWebBrowser browserControl, IBrowser browser, KeyType type, int windowsKeyCode, int nativeKeyCode, CefEventFlags modifiers, bool isSystemKey)
        {
            if (type == KeyType.KeyUp && Enum.IsDefined(typeof(Keys), windowsKeyCode))
            {
                var key = (Keys)windowsKeyCode;
                switch (key)
                {
                    case Keys.F5:
                        browser.Reload(true);
                        break;
                    case Keys.F12:
                        browser.ShowDevTools();
                        break;
                }
            }
            return false;
        }
}
```

Making sure to implement both methods of the IKeyboardHandler

**Reply by wnunezc** [reply]
thanks i will try this.
